// file_reader/index.js
const fs = require('fs');
class FileReader{
    static async readFile(filename){
        let arrayOfFileLineStrings = [];
        let input = fs.createReadStream(filename);
        let readLineObj = require('readline').createInterface({
            input: input, 
            terminal: false
        });

        // Loop through and add to array
        for await (const line of readLineObj) {
            arrayOfFileLineStrings.push(String(line).trim());
        }

        readLineObj.close();
        return arrayOfFileLineStrings;
    }

    static saveFile(filename, arrayOfStrings){
        let writeStream = fs.createWriteStream(filename, {encoding:'utf-8', flag:'w'})
        for (let i = 0; i < arrayOfStrings.length; i++){
            writeStream.write(arrayOfStrings[i] + '\r\n');
        }
        writeStream.end();
    }

    // EXPECTS FORMAT OF Line 1: Col1Name, Col2Name, Line 2-N: {Data_IN_COL1_INDEX_I}, {DATA_IN_COL2_INDEX_I}
    // ASSUMES 1+ LINES
    static async readCSV(fileName){
        let arrayOfLineStrings = await FileReader.readFile(fileName);
        let arrayOfJSONObjects = [];
        let types = arrayOfLineStrings[0].split(",");
        for (let i = 1; i < arrayOfLineStrings.length; i++){
            let newJSONObj = {};
            let line = arrayOfLineStrings[i];
            let lineSplit = line.split(",");
            for (let j = 0; j < types.length; j++){
                newJSONObj[types[j]] = lineSplit[j];
            }
            arrayOfJSONObjects.push(newJSONObj);
        }
        return arrayOfJSONObjects;
    }

}
module.exports = { FileReader };
