// grocery_list/index.js
const GROCERY_ITEM_MODULE = require("grocery_item");
const FILE_READER_MODULE = require("file_reader"); 
const fileReader = FILE_READER_MODULE.FileReader;
// Server-Side Grocery List representation
// TODO: I know I used a json object with a list inside to store items I may just change to a list we'll see
class GroceryList{
    constructor(){
        this._listData = {"items":[]};
    }

    addItem(name){
        this._listData["items"].push(new GROCERY_ITEM_MODULE.GroceryItem(dataText, 0, ""));
    }

    addItemFromJSON(jsonObj){
        this._listData["items"].push(GROCERY_ITEM_MODULE.GroceryItem.fromJSON(jsonObj));
    }

    toDataString(){
        // TODO
    }

    toJSON(){
        let jsonResponse = {}
        jsonResponse["items"] = [];
        for (let item of this._listData["items"]){
            jsonResponse["items"].push(item.toJSON());
        }
        return jsonResponse;
    }

    fromDataJSON(dataJSON){
        // TODO
    }

    getLength(){
        return this._listData["items"].length;
    }

    getList(){
        return this._listData["items"];
    }

    // Assumes element @ index exists
    getByIndex(index){
        return this._listData["items"][index];
    }

    readFromFile(fileName){
        let csvMapList = fileReader.readCSV(fileName);
        for (let jsonObj of csvMapList){
            this.addItemFromJSON(jsonObj);
        }
    }
}
module.exports = { GroceryList }
